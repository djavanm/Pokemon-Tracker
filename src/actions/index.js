export const setPokemon = pokemon => ({
  type: 'SET_POKEMON',
  pokemon
});

export const catchPokemon = pokemon => ({
  type: 'CATCH_POKEMON',
  pokemon
});

export const releasePokemon = id => ({
  type: 'RELEASE_POKEMON',
  id
});
