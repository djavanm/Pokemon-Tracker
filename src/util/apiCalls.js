export const getPokemon = async indentifier => {
  const url = `https://pokeapi.co/api/v2/pokemon/${indentifier}/`;
  const response = await fetch(url);
  if(!response.ok) {
    throw new Error('There was an error getting your pokemon.');
  };
  const pokemon = await response.json();
  return pokemon;
};
