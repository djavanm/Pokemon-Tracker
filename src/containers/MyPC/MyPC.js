import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { FaLaptop } from 'react-icons/fa';
import { getPokemon } from '../../util/apiCalls';
import { setPokemon, addToTeam, removeFromTeam } from '../../actions';
import { Link } from 'react-router-dom';

export class MyPC extends Component {

  showPokemon = async (e, newId) => {
    e.preventDefault();
    const { setPokemon } = this.props;
    const newPokemon = await getPokemon(newId);
    setPokemon(newPokemon);
    console.log(newPokemon);
  };

  addPokemon = pokemon => {
    const { addToTeam } = this.props;
    addToTeam(pokemon);
  }

  removePokemon = pokemon => {
    const { removeFromTeam } = this.props;
    removeFromTeam(pokemon.id)
  }

  render() {
    const { caughtPokemon, teamPokemon, currentPokemon } = this.props;
    let teamSprites = teamPokemon.map((pokemon, index) => {
      return <img className='team sprite' src={`${pokemon.sprite}`} alt={`{pokemon.name}'s sprite`} key={index} onClick={(e) => this.showPokemon(e, pokemon.id)}/>
    });
    let allSprites = caughtPokemon.map((pokemon, index) => {
      let teamClass = teamPokemon.map(pokemon => pokemon.id).includes(pokemon.id) ? 'onTeam' : null;
      return <img className={`sprite ${teamClass}`} src={`${pokemon.sprite}`} alt={`${pokemon.name}'s sprite`} key={index} onClick={(e) => this.showPokemon(e, pokemon.id)}/>
    });
    const addDisabled = teamPokemon.length >= 6 || teamPokemon.map(pokemon => pokemon.id).includes(currentPokemon.id);
    const removeDisabled = !teamPokemon.map(pokemon => pokemon.id).includes(currentPokemon.id);
    return (
      <section className='my-pc-container'>
        <p className='current-team-text'>Curent Team:
        {caughtPokemon.length === 0 && <span className='catch-warning'> Catch Pokemon to add them to your team!</span>}
        {caughtPokemon.length === 0 &&<Link to='/'><button className="team-btn">Catch Pokemon</button></Link>}
        {caughtPokemon.length > 0 && teamPokemon.length === 0 && <span className='catch-warning'> Add some Pokemon to your team!</span>}
        </p>
        <div className='team-container'>{teamSprites}</div>
        {caughtPokemon.length > 0 && <h3 className='mypc-laptop-logo'><FaLaptop color='#01579E' size={30} /></h3>}
        <div className='sprites'>
        {allSprites}
        </div>
        {caughtPokemon.length > 0 && <div className="team-button-container">
        <button className="team-btn" disabled={addDisabled} onClick={() => this.addPokemon(currentPokemon)}> Add To Team </button>
        <button className="team-btn" disabled={removeDisabled} onClick={() => this.removePokemon(currentPokemon)}> Remove From Team </button>
        </div>}
      </section>
    )
  }
}

export const mapStateToProps = state => ({
  currentPokemon: state.currentPokemon,
  caughtPokemon: state.caughtPokemon,
  teamPokemon: state.teamPokemon
});

export const mapDispatchToProps = dispatch => (
  bindActionCreators({ setPokemon, addToTeam, removeFromTeam }, dispatch)
)

export default connect(mapStateToProps, mapDispatchToProps)(MyPC);
