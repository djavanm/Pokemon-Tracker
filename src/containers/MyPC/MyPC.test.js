import React from 'react';
import { shallow } from 'enzyme';
import { MyPC, mapStateToProps, mapDispatchToProps } from './MyPC';
import { setPokemon, addToTeam, removeFromTeam } from '../../actions';
import { FaLaptop } from 'react-icons/fa';
import { getPokemon, getDescription } from '../../util/apiCalls';
jest.mock('../../util/apiCalls');

getDescription.mockImplementation(() => {
  return 'Bulbasaur is cool.';
});

getPokemon.mockImplementation(() => {
  return {
    name: 'Bulbasaur',
    description: 'Bulbasaur is cool.',
    gen: 'Generation 1',
    gif: 'https://projectpokemon.org/images/normal-sprite/bulbasaur.gif',
    height: 7,
    id: 1,
    image: 'https://assets.pokemon.com/assets/cms2/img/pokedex/full/001.png',
    sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png'
  };
});

describe('MyPC', () => {
  let wrapper;
  const mockPokemon = {
    name: 'Bulbasaur',
    description: 'Bulbasaur is cool.',
    gen: 'Generation 1',
    gif: 'https://projectpokemon.org/images/normal-sprite/bulbasaur.gif',
    height: 7,
    id: 1,
    image: 'https://assets.pokemon.com/assets/cms2/img/pokedex/full/001.png',
    sprite: 'https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png'
  };

  const mockCaughtPokemon = [mockPokemon];
  const mockTeamPokemon = [mockPokemon];
  const mockCurrentPokemon =  mockPokemon;
  const mockState = {
    currentPokemon: mockPokemon,
    caughtPokemon: mockCaughtPokemon,
    teamPokemon: mockTeamPokemon
  };


  beforeEach(() => {
    wrapper = shallow(<MyPC
      currentPokemon={mockCurrentPokemon}
      caughtPokemon={mockCaughtPokemon}
      teamPokemon={mockTeamPokemon}
      />)
  });

  it('should match the snapshot given the correct data', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('mapStateToProps should grab the props it needs', () => {
    const expected = mockState;
    const mappedProps = mapStateToProps(mockState);
    expect(mappedProps).toEqual(expected);
  });

  it('it calls dispatch with the setPokemon action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = setPokemon(mockPokemon);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.setPokemon(mockPokemon);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('it calls dispatch with the addPokemon action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = addToTeam(mockPokemon);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.addToTeam(mockPokemon);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('it calls dispatch with the removePokemon action', () => {
    const mockDispatch = jest.fn();
    const actionToDispatch = removeFromTeam(mockPokemon.id);
    const mappedProps = mapDispatchToProps(mockDispatch);
    mappedProps.removeFromTeam(mockPokemon.id);
    expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch);
  });

  it('should get a new pokemon when the show Pokemon method is called', () => {
    const mockEvent = {
      preventDefault: jest.fn()
    };
    wrapper.instance().showPokemon(mockEvent, 1);
    expect(getPokemon).toHaveBeenCalledWith(1);
    wrapper.instance().showPokemon(mockEvent, 'charizard');
    expect(getPokemon).toHaveBeenCalledWith('charizard');
  });  
});
