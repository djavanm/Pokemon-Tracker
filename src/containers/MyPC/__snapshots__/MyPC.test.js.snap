// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MyPC should match the snapshot given the correct data 1`] = `
<section
  className="my-pc-container"
>
  <p
    className="current-team-text"
  >
    Curent Team:
  </p>
  <div
    className="team-container"
  >
    <img
      alt="{pokemon.name}'s sprite"
      className="team sprite"
      key="0"
      onClick={[Function]}
      src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
    />
  </div>
  <h3
    className="mypc-laptop-logo"
  >
    <FaLaptop
      color="#01579E"
      size={30}
    />
  </h3>
  <div
    className="sprites"
  >
    <img
      alt="Bulbasaur's sprite"
      className="sprite onTeam"
      key="0"
      onClick={[Function]}
      src="https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/1.png"
    />
  </div>
  <div
    className="team-button-container"
  >
    <button
      className="team-btn rem"
      disabled={false}
      onClick={[Function]}
    >
       Remove From Team 
    </button>
    <button
      className="team-btn"
      disabled={true}
      onClick={[Function]}
    >
       Add To Team 
    </button>
  </div>
</section>
`;
